examples/gif_paintable/main.rs

```diff
@@ -0,0 +1,239 @@
+// A custom GDK paintable capable of rendering a GIF
+// The paintable makes uses of the awesome image
+// crate to read a gif file and transform it to a Vec<Frame>
+// which are then rendered by the paintable at different snapshots
```

Comment:
From the previous PR [here](https://github.com/gtk-rs/gtk4-rs/pull/152/files#r553270088) (I left unfixed):



> That sounds like a bad idea, especially if the GIF has loops. You can easily use up all memory like this :)
>
> It would be better to decode on demand.

I'm not too familiar with the GTK memory model here but I cant see how this would eat up a ton of memory. The file bytes are read in once and passed by reference to the decoder.

 Is there some cloning going on somewhere that would make this an issue? I can't say I expect folks to load in multi-gigabyte gifs.

Reply:
If you have a GIF with looping, you'd presumably get infinite frames from the decoder

Reply:
Ah I thought you were referring to looping in the UI which is done by indexing the frames. I actually didn't know that gifs had an internal looping mechanism. However, I think this example should prove that the decoder only decodes into the frames a single time: 
```
use image::codecs::gif;
use image::AnimationDecoder;

/// encode a single pixel to a gif that repeats infinitely
fn make_gif() -> Vec<u8> {
    let mut bytes = Vec::new();
    {
        let mut encoder = gif::GifEncoder::new(&mut bytes);
        encoder.set_repeat(gif::Repeat::Infinite).unwrap();
        encoder
            .encode(&[0, 0, 0, 0], 1, 1, image::ColorType::Rgba8)
            .unwrap();
    }
    bytes
}

/// decode a gif into the constituent frames
fn decode_to_frames(bytes: Vec<u8>) -> Vec<image::Frame> {
    let decoder = gif::GifDecoder::new(std::io::Cursor::new(bytes)).unwrap();
    let frames = decoder.into_frames();
    let frames = frames.collect_frames().expect("error decoding gif");
    frames
}

fn main() {
    let bytes = make_gif();
    let frames = decode_to_frames(bytes);
    assert_eq!(frames.len(), 1);
}

```

https://play.rust-lang.org/?version=stable&mode=debug&edition=2018&gist=0418f0575168ea1d956c7044c836e69c

Reply:
That's good, so this handles looping better than the gdkpixbuf animation API where the equivalent code would just use up all your system's memory :)

Reply:
How would the application know from the `Vec<Frame>` that these have to be repeated though?

Reply:
I don't see any methods on the `GifDecoder` [here](https://docs.rs/image/0.23.14/image/codecs/gif/struct.GifDecoder.html), or any methods on the underlying `gif` crate [here](https://docs.rs/gif/0.11.2/gif/struct.Decoder.html) so I believe we are hand-waving away the looping issues. 

The code provided loops infinitely anyway.